### PR TITLE
Added go-conntrack for monitoring http connections

### LINKED ISSUE #2169: Use go-conntrack
https://github.com/mwitkow/go-conntrack gives us the ability to see what http sockets are open, and also adds some metrics. We've had a few issues in the past with leaks so this should be very useful.

### PR DESCRIPTION
**Pull request by slintes**
Hi, I'd like to get more familiar with Prometheus, and this is my first contribution.

With this change you get these metrics:

```
# HELP net_conntrack_dialer_conn_attempted_total Total number of connections attempted by the given dialer a given name.
# TYPE net_conntrack_dialer_conn_attempted_total counter
net_conntrack_dialer_conn_attempted_total{dialer_name="default"} 1
# HELP net_conntrack_dialer_conn_closed_total Total number of connections closed which originated from the dialer of a given name.
# TYPE net_conntrack_dialer_conn_closed_total counter
net_conntrack_dialer_conn_closed_total{dialer_name="default"} 0
# HELP net_conntrack_dialer_conn_established_total Total number of connections successfully established by the given dialer a given name.
# TYPE net_conntrack_dialer_conn_established_total counter
net_conntrack_dialer_conn_established_total{dialer_name="default"} 1
# HELP net_conntrack_dialer_conn_failed_total Total number of connections failed to dial by the dialer a given name.
# TYPE net_conntrack_dialer_conn_failed_total counter
net_conntrack_dialer_conn_failed_total{dialer_name="default",reason="refused"} 0
net_conntrack_dialer_conn_failed_total{dialer_name="default",reason="resolution"} 0
net_conntrack_dialer_conn_failed_total{dialer_name="default",reason="timeout"} 0
net_conntrack_dialer_conn_failed_total{dialer_name="default",reason="unknown"} 0
# HELP net_conntrack_listener_conn_accepted_total Total number of connections opened to the listener of a given name.
# TYPE net_conntrack_listener_conn_accepted_total counter
net_conntrack_listener_conn_accepted_total{listener_name="http"} 10
# HELP net_conntrack_listener_conn_closed_total Total number of connections closed that were made to the listener of a given name.
# TYPE net_conntrack_listener_conn_closed_total counter
net_conntrack_listener_conn_closed_total{listener_name="http"} 8
```

Fixes #2169

### CODE REVIEW

File: web/web.go [hunk 1]
```diff
@@ -348,6 +349,12 @@ func (h *Handler) Run() {
 		ReadTimeout: h.options.ReadTimeout,
 	}
 	listener, err := net.Listen("tcp", h.options.ListenAddress)
+
+	// Monitor incoming connections with conntrack
```
Comment (brian-brazil) [inline web/web.go]: Comments should end in a full stop

File: cmd/prometheus/main.go [hunk 2]
```diff
@@ -154,6 +156,11 @@ func Main() int {
 	webHandler := web.New(&cfg.web)
 	go webHandler.Run()
 
+	// Monitor outgoing connections with conntrack
+	http.DefaultTransport.(*http.Transport).DialContext = conntrack.NewDialContextFunc(
```
Comment (brian-brazil) [inline cmd/prometheus/main.go]: Most connections of interest aren't made with the default transport, look at NewClientFromConfig in util/httputil and at the discovery directory.

**Comment by redbaron** [comment]
A bit of user feedback: `conntrack` is usually associated with Linux kernel connection tracking feature, seeing `net_conntrack` metric names makes it  a bit confusing

**Comment by brian-brazil** [comment]
@redbaron I also had that confusion. That would be something to take up with upstream.

**Comment by slintes** [comment]
It's not possible to customize the metric names, that would indeed mean a change in the go-conntrack library.

**Comment by brian-brazil** [comment]
This looks generally okay, but is going to cause some conflicts with recent changes plus changes to 2.0. To minimise disruption would you mind making this change against the dev-2.0 branch?

**Comment by slintes** [comment]
ok, was just busy with rebasing on master. I see failing tests for the file discovery, is that a know issue at the moment?

```
time="2017-10-05T16:20:51+02:00" level=error msg="Error reading file "fixtures/_test_valid.yml": yaml: did not find expected node content" source="file.go:200"
time="2017-10-05T16:20:51+02:00" level=error msg="Error reading file "fixtures/_test_valid.json": unexpected end of JSON input" source="file.go:200"
time="2017-10-05T16:20:51+02:00" level=error msg="Error reading file "fixtures/_test_valid.json": invalid character ']' looking for beginning of value" source="file.go:200"
time="2017-10-05T16:20:52+02:00" level=error msg="Error reading file "fixtures/_test_valid.json": invalid character ']' looking for beginning of value" source="file.go:200"
time="2017-10-05T16:20:52+02:00" level=error msg="Error reading file "fixtures/_test_invalid_nil.json": unexpected end of JSON input" source="file.go:200"
time="2017-10-05T16:20:52+02:00" level=error msg="Error reading file "fixtures/_test_valid.json": invalid character ']' looking for beginning of value" source="file.go:200"
time="2017-10-05T16:20:52+02:00" level=error msg="Error reading file "fixtures/_test_invalid_nil.json": nil target group item found" source="file.go:200"
time="2017-10-05T16:20:52+02:00" level=error msg="Error reading file "fixtures/_test_valid.json": invalid character ']' looking for beginning of value" source="file.go:200"
time="2017-10-05T16:21:07+02:00" level=error msg="Error reading file "fixtures/_test_invalid_nil.json": invalid character ']' looking for beginning of value" source="file.go:200"
time="2017-10-05T16:21:07+02:00" level=error msg="Error reading file "fixtures/_test_valid.json": invalid character ']' looking for beginning of value" source="file.go:200"
time="2017-10-05T16:21:07+02:00" level=error msg="Error reading file "fixtures/_test_valid.yml": yaml: did not find expected node content" source="file.go:200"
time="2017-10-05T16:21:07+02:00" level=error msg="Error reading file "fixtures/_test_valid.yml": yaml: did not find expected node content" source="file.go:200"
time="2017-10-05T16:21:07+02:00" level=error msg="Error reading file "fixtures/_test_valid.yml": yaml: did not find expected node content" source="file.go:200"
--- FAIL: TestFileSD (16.62s)
	file_test.go:95: Unexpected target groups [], we expected a failure here.
FAIL
FAIL	github.com/prometheus/prometheus/discovery/file	16.634s
```

**Comment by slintes** [comment]
btw: I tried to configure go-conntrack for as many discoveries as possible, but for several discoveries the http client is well hidden in their SDK, so I had no chance to do anything. If they use the `default` transport they will at least be monitored, only without an own label, but if not they won't be visible in the metrics at all. :/

**Comment by brian-brazil** [comment]
> or several discoveries the http client is well hidden in their SDK, so I had no chance to do anything.

That's fine, this will at least let us monitor some of them.

**Comment by brian-brazil** [comment]
> I see failing tests for the file discovery, is that a know issue at the moment?

Tests should be passing on both master and dev-2.0.

**Comment by slintes** [comment]
Ok, moved my changes to the dev-2.0 branch and edited the base branch of this PR. Let's see what happens, never did that before ;)

File: storage/remote/client.go [hunk 3]
```diff
@@ -51,7 +51,7 @@ type clientConfig struct {
 
 // NewClient creates a new Client.
 func NewClient(index int, conf *clientConfig) (*Client, error) {
-	httpClient, err := httputil.NewClientFromConfig(conf.httpClientConfig)
+	httpClient, err := httputil.NewClientFromConfig(conf.httpClientConfig, "storage")
```
Comment (brian-brazil) [inline storage/remote/client.go]: remote_storage would be clearer

File: discovery/consul/consul.go [hunk 4]
```diff
@@ -106,7 +107,13 @@ func NewDiscovery(conf *config.ConsulSDConfig, logger log.Logger) (*Discovery, e
 	if err != nil {
 		return nil, err
 	}
-	transport := &http.Transport{TLSClientConfig: tls}
+	transport := &http.Transport{
+		TLSClientConfig: tls,
+		DialContext: conntrack.NewDialContextFunc(
+			conntrack.DialWithTracing(),
+			conntrack.DialWithName("consul"),
```
Comment (brian-brazil) [inline discovery/consul/consul.go]: consul_sd would be better, as if this ends up being used in another application with other consul use that'll be clearer.

File: storage/remote/client.go (hunk 3 above)
Comment (slintes) [reply storage/remote/client.go]: done

File: discovery/consul/consul.go (hunk 4 above)
Comment (slintes) [reply discovery/consul/consul.go]: done, also for marathon and triton discoveries.

**Comment by brian-brazil** [comment]
Thanks!

**Pull request merged**